mlflow/lightgbm.py

```diff
@@ -182,3 +192,111 @@ def __init__(self, lgb_model):
+    - trained model.
+    Note that the `scikit-learn API`_ is not supported.
+    :param importance_types: importance types to log.
+    """
+    import lightgbm
+
+    @gorilla.patch(lightgbm)
+    def train(*args, **kwargs):
+
+        def record_eval_results(eval_results):
+            """
+            Create a callback function that records evaluation results.
+            """
+            def callback(env):
+                res = {}
+                for data_name, eval_name, value, _ in env.evaluation_result_list:
+                    key = data_name + '-' + eval_name
+                    res[key] = value
+
+                eval_results.append(res)
+            return callback
+
+        if not mlflow.active_run():
+            try_mlflow_log(mlflow.start_run)
+            auto_end_run = True
+        else:
+            auto_end_run = False
+
+        original = gorilla.get_original_attribute(lightgbm, 'train')
+
+        # logging booster params separately via mlflow.log_params to extract key/value pairs
+        # and make it easier to compare them across runs.
+        params = args[0] if len(args) > 0 else kwargs['params']
+        try_mlflow_log(mlflow.log_params, params)
+
+        unlogged_params = ['params', 'train_set', 'valid_sets', 'valid_names', 'fobj', 'feval',
+                           'init_model', 'evals_result', 'learning_rates', 'callbacks']
+
+        log_fn_args_as_params(original, args, kwargs, unlogged_params)
+
+        all_arg_names = inspect.getargspec(original)[0]  # pylint: disable=W1505
+        num_pos_args = len(args)
+
+        # adding a callback that records evaluation results.
+        eval_results = []
+        callbacks_index = all_arg_names.index('callbacks')
+        callback = record_eval_results(eval_results)
+        if num_pos_args >= callbacks_index + 1:
+            tmp_list = list(args)
+            tmp_list[callbacks_index] += [callback]
+            args = tuple(tmp_list)
+        elif 'callbacks' in kwargs and kwargs['callbacks'] is not None:
+            kwargs['callbacks'] += [callback]
+        else:
+            kwargs['callbacks'] = [callback]
+
+        # training model
+        model = original(*args, **kwargs)
+
+        # logging metrics on each iteration.
+        for idx, metrics in enumerate(eval_results):
+            try_mlflow_log(mlflow.log_metrics, metrics, step=idx)
+
+        # If early_stopping_rounds is present, logging metrics at the best iteration
+        # as extra metrics with the max step + 1.
+        early_stopping_index = all_arg_names.index('early_stopping_rounds')
+        early_stopping = (num_pos_args >= early_stopping_index + 1 or
+                          'early_stopping_rounds' in kwargs)
+        if early_stopping:
+            extra_step = len(eval_results)
+            try_mlflow_log(mlflow.log_metric, 'stopped_iteration', len(eval_results))
+            try_mlflow_log(mlflow.log_metric, 'best_iteration', model.best_iteration)
```

Comment:
Just to confirm, it sounds ([from the docs](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.train.html#lightgbm-train)) that `best_iteration` is set as long as the user configures their Booster to use early stopping, even if we end up training for the full number of training epochs & do not stop early (?)

Reply:
Yes, it doesn't matter whether or not the training early stopped.

https://lightgbm.readthedocs.io/en/latest/_modules/lightgbm/callback.html#early_stopping

Reply:
Wrote a simple test code to confirm this.

```python
# code
model = lgb.train(params,
                  train_set,
                  num_boost_round=2,
                  early_stopping_rounds=10,  # setting larger value than num_boost_round
                  valid_sets=[train_set, valid_set],
                  valid_names=['train', 'valid'])
print(model.best_iteration)
print(model.best_score)
```

output

```
[1]     train's multi_logloss: 0        valid's multi_logloss: 0
Training until validation scores don't improve for 10 rounds
[LightGBM] [Warning] Stopped training because there are no more leaves that meet the split requirements
[2]     train's multi_logloss: 0        valid's multi_logloss: 0
Did not meet early stopping. Best iteration is:
[1]     train's multi_logloss: 0        valid's multi_logloss: 0
1
defaultdict(<class 'collections.OrderedDict'>, {'train': OrderedDict([('multi_logloss', 0.0)]), 'valid': OrderedDict([('multi_logloss', 0.0)])})
```

Reply:
Added a comment on this behavior for clarification.